File: Cadeiras/Conceitos de Linguagens de Computacao/Jodo da Vida/jogoDaVida.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#define tam 100

void criaJogo(int tamanhoTabuleiro,int percentual,int nGeracoes);
int main(){
	clock_t c2, c1; /* variáveis que contam ciclos do processador */
	double tempo;
	c1 = clock();
	int tamanho,percentual,nGeracoes;
	printf("Tamanho do Tabuleiro:");
	scanf("%d",&tamanho);
	printf("Percentual de geracoes:");
	scanf("%d",&percentual);
	printf("Quantas geracoes:");
	scanf("%d",&nGeracoes);
	criaJogo(tamanho,percentual,nGeracoes);
	c2 = clock();
	tempo = (c2 - c1)/(double)CLOCKS_PER_SEC;
	printf ("%lf", tempo);
	return 0;
}
void criaJogo(int tamanhoTabuleiro,int percentual,int nGeracoes){
	int tabuleiro[tam][tam],tabuleiroAux[tam][tam],i,j,AreaTabuleiro,geracoes,contador=0,contadorVizinhos=0,auxiliar,contadorFim=0;
	for (i=0;i<tamanhoTabuleiro;i++){//zera o tabuleiro para efetuar comparação na randomização
		for (j=0;j<tamanhoTabuleiro;j++){
			tabuleiro[i][j]=0;
		}
	}
	AreaTabuleiro=tamanhoTabuleiro*tamanhoTabuleiro;
	geracoes=(AreaTabuleiro*percentual)/100;
	while(contador!=geracoes){//para de randomizar quando tiver acessado todas as posiçoes do tabuleiro
		i = rand()%tamanhoTabuleiro;//randomiza o i
		j = rand()%tamanhoTabuleiro;//randomiza o j
		if(tabuleiro[i][j]==0){//toda vez que uma posiçao sorteada no tabuleiro for igual a zero ele troca por 1
			tabuleiro[i][j]=1;
			contador++;//conta quantas vezes ele acessou o tabuleiro
		}
	}
	printf("\nOrigem das geracoes:\n\n");
	for (i=0;i<tamanhoTabuleiro;i++){			//imprime o tabuleiro inicial,após randomização
		for (j=0;j<tamanhoTabuleiro;j++){
			printf("%d",tabuleiro[i][j]);
		}
		printf("\n");
	}
	auxiliar=nGeracoes;//avisa em qual geração esta
	while(auxiliar>0){
		for (i=0;i<tamanhoTabuleiro;i++)
			for (j=0;j<tamanhoTabuleiro;j++){
			tabuleiroAux[i][j]=tabuleiro[i][j];
		}
		for (i=0; i<tamanhoTabuleiro; i++)
			for (j=0; j<tamanhoTabuleiro; j++) {
			contadorVizinhos=0;
			if(i==0&&j==0){	//verifica canto superior esquerdo do tabuleiro
				if(tabuleiro[i+1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j+1]==1)
					contadorVizinhos++;
				if(tabuleiro[i][j+1]==1)
					contadorVizinhos++;
				}
			else if(i==tamanhoTabuleiro-1&&j==tamanhoTabuleiro-1){	//verifica canto inferior direito do tabuleiro
				if(tabuleiro[i-1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j-1]==1)
					contadorVizinhos++;
				}
			else if (i==0&&j==tamanhoTabuleiro-1){ //verifica canto superio direito do tabuleiro 		
				if(tabuleiro[i+1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j-1]==1)
					contadorVizinhos++;
				}
			else if ((i==0&&(j>0&&j<tamanhoTabuleiro-1))) { //verifica parte de cima do tabuleiro
				if(tabuleiro[i+1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i][j+1]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j+1]==1)
					contadorVizinhos++;
				}
			else if(i==tamanhoTabuleiro-1&&j==0){ //verifica canto inferior esquerdo do tabuleiro
				if(tabuleiro[i-1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i][j+1]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j+1]==1)
					contadorVizinhos++;
				}
			else if(i==tamanhoTabuleiro-1&&j>0&&j<tamanhoTabuleiro-1){ //verifica parte de baixo do tabuleiro
				if(tabuleiro[i][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i][j+1]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j+1]==1)
					contadorVizinhos++;
				}
			else if((i>0&&i<tamanhoTabuleiro-1)&&j==0){ //verifica lado esquerdo do tabuleiro
				if(tabuleiro[i-1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i][j+1]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j+1]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j+1]==1)
					contadorVizinhos++;
				}				
			else if(i>0&&i<tamanhoTabuleiro&&j==tamanhoTabuleiro-1){ //verifica lado direito do tabuleiro
				if(tabuleiro[i][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j-1]==1)
					contadorVizinhos++;
				}
			else if((i>0&&i<tamanhoTabuleiro-1)&&(j>0&&j<tamanhoTabuleiro-1)){ //verifica centro do tabuleiro
				if(tabuleiro[i][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j]==1)
					contadorVizinhos++;
				if(tabuleiro[i][j+1]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j-1]==1)
					contadorVizinhos++;
				if(tabuleiro[i-1][j+1]==1)
					contadorVizinhos++;
				if(tabuleiro[i+1][j+1]==1)
					contadorVizinhos++;
			}
			if(tabuleiro[i][j]==1&&((contadorVizinhos<2)||(contadorVizinhos>3)))
				tabuleiroAux[i][j]=0;
			if(tabuleiro[i][j]==0&&(contadorVizinhos==3))
				tabuleiroAux[i][j]=1;
		    if((tabuleiro[i][j]==1)&&((contadorVizinhos==2)||(contadorVizinhos==3)))
				tabuleiroAux[i][j]=tabuleiro[i][j];
			}
		contadorFim=0;
		for (i=0;i<tamanhoTabuleiro;i++)
			for (j=0;j<tamanhoTabuleiro;j++){
				if(tabuleiroAux[i][j]==tabuleiro[i][j])
					contadorFim++;
				}
		for (i=0;i<tamanhoTabuleiro;i++)
			for (j=0;j<tamanhoTabuleiro;j++){
				tabuleiro[i][j]=tabuleiroAux[i][j];
			}
		if (contadorFim==AreaTabuleiro){
			printf("\n*Nao e necessario imprimir mais geracoes, as demais serao iguais!\n\n");
			return;
		}
		printf("\nGeracao:%d\n\n",(nGeracoes-auxiliar)+1);//imprime em qual geracao esta
		for (i=0;i<tamanhoTabuleiro;i++){//imprime as geracoes
			for (j=0;j<tamanhoTabuleiro;j++){
				printf("%d",tabuleiroAux[i][j]);
			}
			printf("\n");
		}
	auxiliar--;
	}
	
}

File: Cadeiras/Conceitos de Linguagens de Computacao/Jodo da Vida/jogoDaVida.f90
program jogoDaVida!inicia o programa
integer:: tamanho,percentual,nGeracoes
real:: time_begin,time_end
call cpu_time ( time_begin )
write(*,*)'Tamanho Do Tabuleiro:'
read (*,*) tamanho
write(*,*)'Percentual de Geracoes:'
read (*,*) percentual
write(*,*)'Numero de Geracoes:'
read (*,*) nGeracoes
call criaJogo(tamanho,percentual,nGeracoes)!passa pra subroutina os dados sobre o jogo
call cpu_time (time_end)
print *, 'tempo:', time_end - time_begin, 's'
end
subroutine criaJogo(tamanhoTabuleiro,percentual,nGeracoes) 
integer:: tamanhoTabuleiro,percentual,nGeracoes,i,j,AreaTabuleiro,geracoes,contador=0,auxiliar,contadorFim
integer:: tabuleiro(50,50),tabuleiroAux(50,50)
real:: randI,randJ
do i=1, tamanhoTabuleiro!atribui 0 a todas as posiçoes do tabuleiro para posterior randomizacoo das posicoes das celulas
do j=1, tamanhoTabuleiro
tabuleiro(i,j)=0
end do
end do
AreaTabuleiro=tamanhoTabuleiro*tamanhoTabuleiro
geracoes=(AreaTabuleiro*percentual)/100
do while(contador/=geracoes)
randI=rand()
randJ=rand()
i=int(randI*tamanhoTabuleiro+1)!converte o numero real randomizado em um inteiro
j=int(randJ*tamanhoTabuleiro+1)
if(tabuleiro(i,j)==0)then
tabuleiro(i,j)=1
contador=contador+1
end if
end do
write (*,*)'Origem das Geracoes:'
do i=1,tamanhoTabuleiro
do j=1,tamanhoTabuleiro
write(*,"(A)",advance="no"),tabuleiro(i,j)+48!é somado 48, pois atraves desse metodo é impresso na tela como char a matriz
tabuleiroAux(i,j)=tabuleiro(i,j)!atribui os valores do tabuleiro para tabuleiroAux
enddo
print*,""!pula a linha quando j for igual ao tamanho do tabuleiro
enddo
auxiliar=nGeracoes
do while(auxiliar>0)
do i=1,tamanhoTabuleiro
do j=1,tamanhoTabuleiro
contadorVizinhos=0
!verifica canto superior esquerdo
if (i==1.AND.j==1)then
if(tabuleiro(i+1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
!verifica canto inferior direito do tabuleiro
else if(i==tamanhoTabuleiro.AND.j==tamanhoTabuleiro)then
if(tabuleiro(i-1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
!verifica canto superio direito do tabuleiro 		
else if(i==1.AND.j==tamanhoTabuleiro)then 
if(tabuleiro(i+1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
!verifica parte de cima do tabuleiro
else if ((i==1.AND.(j>1.AND.j<tamanhoTabuleiro)))then
if(tabuleiro(i+1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
!verifica canto inferior esquerdo do tabuleiro
else if(i==tamanhoTabuleiro.AND.j==1)then
if(tabuleiro(i-1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
!verifica parte de baixo do tabuleiro
else if(i==tamanhoTabuleiro.AND.j>1.AND.j<tamanhoTabuleiro)then
if(tabuleiro(i,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
!verifica lado esquerdo do tabuleiro
else if((i>1.AND.i<tamanhoTabuleiro).AND.j==0)then
if(tabuleiro(i-1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
!verifica lado direito do tabuleiro
else if(i>1.AND.i<tamanhoTabuleiro.AND.j==tamanhoTabuleiro)then
if(tabuleiro(i,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
!verifica centro do tabuleiro
else if((i>1.AND.i<tamanhoTabuleiro).AND.(j>1.AND.j<tamanhoTabuleiro))then
if(tabuleiro(i,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j-1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i-1,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
if(tabuleiro(i+1,j+1)==1)then
contadorVizinhos=contadorVizinhos+1
end if
end if
!regras do jogo
if(tabuleiro(i,j)==1.AND.((contadorVizinhos<2).OR.(contadorVizinhos>3)))then
tabuleiroAux(i,j)=0
end if
if(tabuleiro(i,j)==0.AND.(contadorVizinhos==3))then
tabuleiroAux(i,j)=1
end if
if((tabuleiro(i,j)==1).AND.((contadorVizinhos==2).OR.(contadorVizinhos==3)))then
tabuleiroAux(i,j)=tabuleiro(i,j)
end if
end do
end do
contadorFim=0
do i=1, tamanhoTabuleiro
do j=1, tamanhoTabuleiro
if(tabuleiroAux(i,j)==tabuleiro(i,j))then
contadorFim=contadorFim+1
end if
end do
end do
do i=1, tamanhoTabuleiro
do j=1, tamanhoTabuleiro
tabuleiro(i,j)=tabuleiroAux(i,j)
end do
end do
if (contadorFim==AreaTabuleiro)then
write (*,*)'*Nao e necessario imprimir mais geracoes, as demais serao iguais!'
stop
end if
write (*,*)'Geracao:',(nGeracoes-auxiliar)+1 
do i=1,tamanhoTabuleiro
do j=1,tamanhoTabuleiro
write(*,"(A)",advance="no"),tabuleiroAux(i,j)+48!é somado 48, pois atraves desse metodo é impresso na tela como char a matriz
end do
print*,""!pula a linha quando j for igual ao tamanho do tabuleiro
end do
auxiliar=auxiliar-1
end do
end subroutine


